pkg/endpoint/policy.go

```diff
@@ -408,59 +409,61 @@ func (e *Endpoint) updateRegenerationStatistics(context *regenerationContext, er
 func (e *Endpoint) Regenerate(owner Owner, regenMetadata *ExternalRegenerationMetadata) <-chan bool {
 	done := make(chan bool, 1)
 
-	go func() {
-		var buildSuccess bool
+	regenContext := regenMetadata.toRegenerationContext()
 
-		regenContext := regenMetadata.toRegenerationContext()
+	epEvent := eventqueue.NewEvent(&EndpointRegenerationEvent{
+		owner:        owner,
+		regenContext: regenContext,
+		ep:           e,
+	})
 
-		defer func() {
-			done <- buildSuccess
-			close(done)
-		}()
+	// This may block if the Endpoint's EventQueue is full.
+	resChan := e.EventQueue.Enqueue(epEvent)
 
-		err := e.RLockAlive()
-		if err != nil {
-			e.LogDisconnectedMutexAction(err, "before regeneration")
-			return
-		}
-		e.RUnlock()
-
-		// We should only queue the request after we use all the endpoint's
-		// lock/unlock. Otherwise this can get a deadlock if the endpoint is
-		// being deleted at the same time. More info PR-1777.
-		doneFunc := owner.QueueEndpointBuild(uint64(e.ID))
-		if doneFunc != nil {
-			e.getLogger().Debug("Dequeued endpoint from build queue")
-
-			regenContext.DoneFunc = doneFunc
-			err := e.regenerate(owner, regenContext)
-			doneFunc() // in case not called already
-
-			repr, reprerr := monitorAPI.EndpointRegenRepr(e, err)
-			if reprerr != nil {
-				e.getLogger().WithError(reprerr).Warn("Notifying monitor about endpoint regeneration failed")
-			}
+	go func() {
 
-			if err != nil {
-				buildSuccess = false
-				if reprerr == nil && !option.Config.DryMode {
-					owner.SendNotification(monitorAPI.AgentNotifyEndpointRegenerateFail, repr)
-				}
+		var buildSuccess bool
+		var regenError error
+
+		select {
+		case result, ok := <-resChan:
+			if ok {
+				regenResult := result.(*EndpointRegenerationResult)
+				regenError = regenResult.err
+
+				// Build was successful whether regeneration errored out of not.
```

Comment:
nit: Comment in english of not?

Reply:
Also, even with fuzzing meanings into this comment I don't follow how it relates to the code below. The comment seems more confusing being there than if it were omitted.

Reply:
Will remove it in next upload.